### PR TITLE
RUMM-3014: Extract OkHttp instrumentation into dedicated module

### PR DESCRIPTION
**Pull request by 0xnm**
### What does this PR do?

This change does the extraction of OKHttp instrumentation code into the dedicated `dd-sdk-android-okhttp` module.

To support cross-module communication the following things are done:

* Made some utility methods public
* `FirstPartyHostHeaderTypeResolver` is now split into immutable interface and mutable implementation. The former is exposed as the `SdkCore#firstPartyHostResolver` property, because it should be shared by the core with 2 external modules - `okhttp` and `rum` (not extracted yet).
* Some methods of internal `AdvancedRumMonitor` are moved to the public interface `AdvancedNetworkRumMonitor`, because they are accessed by the `okhttp` module. This is done to avoid having the whole `AdvancedRumMonitor` and related types as public.
* Nightly tests coverage script is updated to account for the `apiSurface` files in different modules.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

**Review by xgouchet — APPROVED** [review]

File: dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/AdvancedNetworkRumMonitor.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+package com.datadog.android.rum.internal.monitor
+
+import com.datadog.android.rum.internal.domain.event.ResourceTiming
+
+/**
+ * FOR INTERNAL USAGE ONLY.
+ */
+@SuppressWarnings("UndocumentedPublicFunction")
+interface AdvancedNetworkRumMonitor {
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/AdvancedNetworkRumMonitor.kt:15]: Great idea

File: tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt [hunk 2]
```diff
@@ -62,7 +60,15 @@ class TestConfigurationExtension :
...
-        context?.callTestConfigurations { forge -> tearDown(forge) }
+        // reverse here is needed, because of the following example:
+        // say there is test class A, and test class B : A.
+        // both declare InternalLoggerTestExtension, which saves original logger in setUp
+        // and restores it during tearDown.
+        // that means that during setup we will save original logger in B, and then mocked
+        // logger in A. But during tear down we will restore first original logger in B, and then
+        // mocked logger in A if we don't reverse the call order. So reversing the order to follow
+        // the same call sequence as JUnit is doing.
+        context?.callTestConfigurations(reverseCallOrder = true) { forge -> tearDown(forge) }
```
Comment (mariusc83) [inline tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt:71]: good catch

File: tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt [hunk 3]
```diff
@@ -107,5 +122,23 @@ class TestConfigurationExtension :
         return ForgeExtension.getForge(this)
     }
 
+    private fun collectProviderMethods(clazz: Class<*>, accumulator: MutableList<Method>) {
```
Comment (mariusc83) [inline tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt:125]: again good catch here

**Review by mariusc83 — APPROVED** [review]
LGTM
Just one comment regarding the API surface config for session replay module but not blocking.

File: instrumentation-tools/src/constants.py [hunk 4]
```diff
@@ -2,7 +2,14 @@
 #  This product includes software developed at Datadog (https://www.datadoghq.com/).
 #  Copyright 2016-Present Datadog, Inc.
 
-API_SURFACE_PATH = "dd-sdk-android/apiSurface"
+API_SURFACE_PATHS = [
+    "dd-sdk-android/apiSurface",
```
[the remark this answers is not in the document]
Comment (0xnm) [reply instrumentation-tools/src/constants.py:6]: yes, it is there, few lines below :)

**Pull request merged**